From 498f94d9f2cb724dcdeb12beeb095d560314b41a Mon Sep 17 00:00:00 2001
From: boogie <boogiepop@gmx.com>
Date: Sat, 13 Jan 2024 19:39:52 +0100
Subject: [PATCH] windowing/gbm: Dynamic plane selection

This commit allows kodi to select gui and video planes according to
format and modifier combination and sorts to gui plane on top of video
plane if zpos is supported and not immutable by the planes.
---
 .../HwDecRender/RendererDRMPRIME.cpp          |  4 +-
 xbmc/windowing/gbm/drm/DRMAtomic.cpp          | 43 +++++++++++++++++++
 xbmc/windowing/gbm/drm/DRMAtomic.h            |  1 +
 xbmc/windowing/gbm/drm/DRMObject.cpp          | 30 +++++++++++++
 xbmc/windowing/gbm/drm/DRMObject.h            |  4 ++
 xbmc/windowing/gbm/drm/DRMUtils.cpp           | 21 +++++----
 xbmc/windowing/gbm/drm/DRMUtils.h             |  5 ++-
 7 files changed, 96 insertions(+), 12 deletions(-)

diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp
index a07b4f744048e..04d42c09828b0 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp
@@ -77,9 +77,11 @@ CBaseRenderer* CRendererDRMPRIME::Create(CVideoBuffer* buffer)
     if (!plane)
       return nullptr;
 
-    if (!plane->SupportsFormatAndModifier(format, modifier))
+    if (!drm->FindPlanes(format, modifier))
       return nullptr;
 
+    drm->SortPlanes();
+
     return new CRendererDRMPRIME();
   }
 
diff --git a/xbmc/windowing/gbm/drm/DRMAtomic.cpp b/xbmc/windowing/gbm/drm/DRMAtomic.cpp
index 5d61a699d4227..04f4c57c42b45 100644
--- a/xbmc/windowing/gbm/drm/DRMAtomic.cpp
+++ b/xbmc/windowing/gbm/drm/DRMAtomic.cpp
@@ -63,6 +63,49 @@ bool CDRMAtomic::SetScalingFilter(CDRMObject* object, const char* name, const ch
   return true;
 }
 
+void CDRMAtomic::SortPlanes()
+{
+  bool supports_zpos = m_gui_plane->SupportsProperty("zpos");
+  bool zpos_immutable = supports_zpos && m_gui_plane->IsPropertyImmutable("zpos").value();
+
+  auto crtc_offset = std::distance(
+      m_crtcs.begin(),
+      std::find_if(m_crtcs.begin(), m_crtcs.end(),
+                   [this](auto& crtc) { return crtc->GetCrtcId() == m_crtc->GetCrtcId(); }));
+
+  // Disable unused planes after planes are re-selected in the active crtc
+  for (auto& plane : m_planes)
+  {
+    if (!(plane.get()->GetPossibleCrtcs() & (1 << crtc_offset)))
+      continue;
+
+    if ((m_video_plane == nullptr || m_video_plane->GetId() != plane.get()->GetId()) &&
+        (m_gui_plane == nullptr || m_gui_plane->GetId() != plane.get()->GetId()))
+    {
+      CLog::Log(LOGDEBUG, "CDRMUtils::{} - disabled plane {}", __FUNCTION__,
+                plane.get()->GetPlaneId());
+      AddProperty(plane.get(), "FB_ID", 0);
+      AddProperty(plane.get(), "CRTC_ID", 0);
+    }
+  }
+
+  if (!supports_zpos || zpos_immutable)
+    return;
+
+  // re-sort the video and gui planes
+  std::optional<uint64_t*> limits = m_gui_plane->GetRangePropertyLimits("zpos");
+
+  if (!limits)
+    return;
+
+  m_gui_plane->SetProperty("zpos", limits.value()[1]);
+  m_video_plane->SetProperty("zpos", limits.value()[0]);
+  CLog::Log(LOGDEBUG, "CDRMUtils::{} - gui plane id,zpos: {}, {}", __FUNCTION__,
+            m_gui_plane->GetId(), limits.value()[0] + 1);
+  CLog::Log(LOGDEBUG, "CDRMUtils::{} - video plane id,zpos: {}, {}", __FUNCTION__,
+            m_video_plane->GetId(), limits.value()[0]);
+}
+
 void CDRMAtomic::DrmAtomicCommit(int fb_id, int flags, bool rendered, bool videoLayer)
 {
   uint32_t blob_id;
diff --git a/xbmc/windowing/gbm/drm/DRMAtomic.h b/xbmc/windowing/gbm/drm/DRMAtomic.h
index ca2cd9a1d0c9b..50fb44f2dd486 100644
--- a/xbmc/windowing/gbm/drm/DRMAtomic.h
+++ b/xbmc/windowing/gbm/drm/DRMAtomic.h
@@ -33,6 +33,7 @@ class CDRMAtomic : public CDRMUtils
   bool InitDrm() override;
   void DestroyDrm() override;
   bool AddProperty(CDRMObject* object, const char* name, uint64_t value);
+  void SortPlanes();
 
   bool DisplayHardwareScalingEnabled();
 
diff --git a/xbmc/windowing/gbm/drm/DRMObject.cpp b/xbmc/windowing/gbm/drm/DRMObject.cpp
index 599bb61cd0234..43f95335cb1ab 100644
--- a/xbmc/windowing/gbm/drm/DRMObject.cpp
+++ b/xbmc/windowing/gbm/drm/DRMObject.cpp
@@ -106,6 +106,25 @@ std::tuple<bool, uint64_t> CDRMObject::GetPropertyValue(const std::string& name,
   return std::make_tuple(false, 0);
 }
 
+std::optional<uint64_t*> CDRMObject::GetRangePropertyLimits(std::string_view name)
+{
+  auto property = std::find_if(m_propsInfo.begin(), m_propsInfo.end(),
+                               [&name](const auto& prop) { return prop->name == name; });
+
+  if (property == m_propsInfo.end())
+    return {};
+
+  auto prop = property->get();
+
+  if (!static_cast<bool>(drm_property_type_is(prop, DRM_MODE_PROP_RANGE)))
+    return {};
+
+  if (prop->count_values != 2)
+    return {};
+
+  return prop->values;
+}
+
 bool CDRMObject::SetProperty(const std::string& name, uint64_t value)
 {
   auto property = std::find_if(m_propsInfo.begin(), m_propsInfo.end(),
@@ -131,3 +150,14 @@ bool CDRMObject::SupportsProperty(const std::string& name)
 
   return false;
 }
+
+std::optional<bool> CDRMObject::IsPropertyImmutable(const std::string& name)
+{
+  auto property = std::find_if(m_propsInfo.begin(), m_propsInfo.end(),
+                               [&name](const auto& prop) { return prop->name == name; });
+
+  if (property == m_propsInfo.end())
+    return {};
+
+  return static_cast<bool>(drm_property_type_is(property->get(), DRM_MODE_PROP_IMMUTABLE));
+}
diff --git a/xbmc/windowing/gbm/drm/DRMObject.h b/xbmc/windowing/gbm/drm/DRMObject.h
index e2ae32652cae1..88e11c2b30c97 100644
--- a/xbmc/windowing/gbm/drm/DRMObject.h
+++ b/xbmc/windowing/gbm/drm/DRMObject.h
@@ -11,6 +11,8 @@
 #include <cstddef>
 #include <cstdint>
 #include <memory>
+#include <optional>
+#include <string_view>
 #include <vector>
 
 #include <xf86drmMode.h>
@@ -39,6 +41,8 @@ class CDRMObject
 
   bool SetProperty(const std::string& name, uint64_t value);
   bool SupportsProperty(const std::string& name);
+  std::optional<bool> IsPropertyImmutable(const std::string& name);
+  std::optional<uint64_t*> GetRangePropertyLimits(std::string_view name);
 
 protected:
   explicit CDRMObject(int fd);
diff --git a/xbmc/windowing/gbm/drm/DRMUtils.cpp b/xbmc/windowing/gbm/drm/DRMUtils.cpp
index 6b61403a339cc..ccc28f937142d 100644
--- a/xbmc/windowing/gbm/drm/DRMUtils.cpp
+++ b/xbmc/windowing/gbm/drm/DRMUtils.cpp
@@ -181,20 +181,23 @@ bool CDRMUtils::FindPreferredMode()
   return true;
 }
 
-bool CDRMUtils::FindPlanes()
+bool CDRMUtils::FindPlanes(uint32_t format, uint64_t modifier)
 {
   for (size_t i = 0; i < m_crtcs.size(); i++)
   {
     if (!(m_encoder->GetPossibleCrtcs() & (1 << i)))
       continue;
 
-    auto videoPlane = std::find_if(m_planes.begin(), m_planes.end(), [&i](auto& plane) {
-      if (plane->GetPossibleCrtcs() & (1 << i))
-      {
-        return plane->SupportsFormat(DRM_FORMAT_NV12);
-      }
-      return false;
-    });
+    auto videoPlane =
+        std::find_if(m_planes.begin(), m_planes.end(),
+                     [&i, &format, &modifier](auto& plane)
+                     {
+                       if (plane->GetPossibleCrtcs() & (1 << i))
+                       {
+                         return plane->SupportsFormatAndModifier(format, modifier);
+                       }
+                       return false;
+                     });
 
     uint32_t videoPlaneId{0};
 
@@ -467,7 +470,7 @@ bool CDRMUtils::InitDrm()
   if (!FindCrtc())
     return false;
 
-  if (!FindPlanes())
+  if (!FindPlanes(DRM_FORMAT_NV12, DRM_FORMAT_MOD_LINEAR))
     return false;
 
   if (!FindPreferredMode())
diff --git a/xbmc/windowing/gbm/drm/DRMUtils.h b/xbmc/windowing/gbm/drm/DRMUtils.h
index 5327e35570467..9c01e54fb0806 100644
--- a/xbmc/windowing/gbm/drm/DRMUtils.h
+++ b/xbmc/windowing/gbm/drm/DRMUtils.h
@@ -62,6 +62,8 @@ class CDRMUtils
   static uint32_t FourCCWithAlpha(uint32_t fourcc);
   static uint32_t FourCCWithoutAlpha(uint32_t fourcc);
 
+  bool FindPlanes(uint32_t format, uint64_t modifier);
+
 protected:
   bool OpenDrm(bool needConnector);
   drm_fb* DrmFbGetFromBo(struct gbm_bo *bo);
@@ -78,13 +80,13 @@ class CDRMUtils
   int m_width = 0;
   int m_height = 0;
 
+  std::vector<std::unique_ptr<CDRMCrtc>> m_crtcs;
   std::vector<std::unique_ptr<CDRMPlane>> m_planes;
 
 private:
   bool FindConnector();
   bool FindEncoder();
   bool FindCrtc();
-  bool FindPlanes();
   bool FindPreferredMode();
   bool RestoreOriginalMode();
   RESOLUTION_INFO GetResolutionInfo(drmModeModeInfoPtr mode);
@@ -95,7 +97,6 @@ class CDRMUtils
 
   std::vector<std::unique_ptr<CDRMConnector>> m_connectors;
   std::vector<std::unique_ptr<CDRMEncoder>> m_encoders;
-  std::vector<std::unique_ptr<CDRMCrtc>> m_crtcs;
 };
 
 }

